File: examples/TestIMU.cpp
#include "Motherboard.hpp"
#include "MotherboardAliases.hpp"

#include <chrono>
#include <thread>

int main()
{
  Roki::Motherboard mb;
  Roki::TTYConfig serviceConfig;

  serviceConfig.Port = "/dev/ttyAMA2";
  serviceConfig.Baudrate = 1250000;
  serviceConfig.ParityBit = true;
  serviceConfig.Stopbits = Roki::TTYConfig::StopbitsCount::One;
  serviceConfig.Timeout = 2;

  if (!mb.Configure(serviceConfig))
  {
    std::cout << mb.GetError() << std::endl;
    exit(1);
  }

  Roki::IMUInfo info;
  Roki::IMUFrame frame;

  uint8_t fr[5]{0, 1, 2, 3, 4};
  int i = -1;
  int j = 0;

  while (false)
  {
    if (j++ == 5)
      if (!mb.ResetIMUCounter())
      {
        std::cout << mb.GetError() << std::endl;
        std::this_thread::sleep_for(std::chrono::seconds(1));
        continue;
      }

    if (!mb.GetIMUInfo(info))
    {
      std::cout << mb.GetError() << std::endl;
      std::this_thread::sleep_for(std::chrono::seconds(1));
      continue;
    }

    std::cout << "Imu info: " << std::endl;

    std::cout << "  First:  " << info.First << std::endl;
    std::cout << "  NumAv:  " << info.NumAv << std::endl;

    std::cout << std::endl;
  }

  while (true)
  {
    i = (i + 1) % 500;
    if (!mb.GetOrientation(frame))
    {
      std::cout << mb.GetError() << std::endl;
      std::this_thread::sleep_for(std::chrono::seconds(1));
      continue;
    }

    std::cout << "Quaternion: " << std::endl;
    std::cout << "  x:  " << frame.Orientation.X << std::endl;
    std::cout << "  y:  " << frame.Orientation.Y << std::endl;
    std::cout << "  z:  " << frame.Orientation.Z << std::endl;
    std::cout << "  w:  " << frame.Orientation.W << std::endl;
    // std::cout << "  Acc: " << frame.Orientation.Accuracy << std::endl;

    std::cout << "Timestamp: " << std::endl;
    std::cout << "  s:  " << frame.Timestamp.TimeS << std::endl;
    std::cout << "  ns: " << frame.Timestamp.TimeNS << std::endl;

    std::cout << "Sensor ID: " << +frame.SensorID << std::endl;
    std::this_thread::sleep_for(std::chrono::milliseconds(300));
  }
  return 0;
}

File: examples/scripts/TestImu.py
import sys
import time

import Roki

mb = Roki.Motherboard()

conf = Roki.TTYConfig()

conf.Port = "/dev/ttyAMA2"
conf.Baudrate = 1250000
conf.Stopbits = Roki.Stopbits.One
conf.ParityBit = True
conf.Timeout = 2

mb.Configure(conf)
#mb.ResetIMUCounter()

while True:
    fr = mb.GetCurrentOrientation()
    #fr = mb.GetOrientationBySeq(150)

    print("Quaternion: ")
    print("  X:  " + str(fr.Orientation.X))
    print("  Y:  " + str(fr.Orientation.Y))
    print("  Z:  " + str(fr.Orientation.Z))
    print("  W:  " + str(fr.Orientation.W))
    print("Timestamp: ")
    print("  S:  " + str(fr.Timestamp.TimeS))
    print("  NS: " + str(fr.Timestamp.TimeNS))
    print("SensorID: " + str(fr.SensorID))
    print("\n")

    info = mb.GetIMUInfo()

    print("Imu info: ")
    print("  First: " + str(info.First))
    print("  NumAv: " + str(info.NumAv))
    print("  MaxFr: " + str(info.MaxFrames))
    print("\n")

    time.sleep(0.5)

